native/dgehrd.go

```diff
@@ -0,0 +1,178 @@
+//  [     v1  v2  h   h   h   h ]
+//  [     v1  v2  v3  h   h   h ]
+//  [                         a ]
+// where a denotes an element of the original matrix A, h denotes a
+// modified element of the upper Hessenberg matrix H, and vi denotes an
+// element of the vector defining H_i.
+//
+// ilo and ihi determine the block of A that will be reduced to upper Hessenberg
+// form. It must hold that 0 <= ilo <= ihi < n if n > 0, and ilo == 0 and ihi ==
+// -1 if n == 0, otherwise Dgehrd will panic.
+//
+// On return, tau will contain the scalar factors of the elementary reflectors.
+// Elements tau[:ilo] and tau[ihi:] of tau will be set to zero. It must have
+// length equal to n-1 if n > 1, otherwise Dgehrd will panic.
+//
+// work must have length at least lwork, otherwise Dgehrd will panic, and lwork
+// must be at least max(1,n). On return, work[0] contains the optimal value of lwork.
+//
+// If lwork is -1, instead of performing Dgehrd, the function only calculates
+// the optimal length of work.
+func (impl Implementation) Dgehrd(n, ilo, ihi int, a []float64, lda int, tau, work []float64, lwork int) {
+	checkMatrix(n, n, a, lda)
+	switch {
+	case ilo < 0 || max(0, n-1) < ilo:
+		panic(badIlo)
+	case ihi < min(ilo, n-1) || n <= ihi:
+		panic(badIhi)
+	case len(work) < lwork:
+		panic(shortWork)
+	case lwork < max(1, n) && lwork != -1:
+		panic(badWork)
+	case n > 0 && len(tau) != n-1:
+		panic(badTau)
+	}
+
+	const (
+		nbmax = 64
+		ldt   = nbmax + 1
+		tsize = ldt * nbmax
+	)
+	// Compute the workspace requirements.
+	nb := min(nbmax, impl.Ilaenv(1, "DGEHRD", " ", n, ilo, ihi, -1))
+	lwkopt := n*nb + tsize
+	work[0] = float64(lwkopt)
+	if lwork == -1 {
+		return
+	}
+
+	// Set tau[:ilo] and tau[ihi:] to zero.
+	for i := 0; i < ilo; i++ {
+		tau[i] = 0
+	}
+	for i := ihi; i < n-1; i++ {
+		tau[i] = 0
+	}
+
+	// Quick return if possible.
+	nh := ihi - ilo + 1
+	if nh <= 1 {
+		work[0] = 1
+		return
+	}
+
+	// Determine the block size.
+	nbmin := 2
+	var nx int
+	if 1 < nb && nb < nh {
+		// Determine when to cross over from blocked to unblocked code
+		// (last block is always handled by unblocked code).
+		nx = max(nb, impl.Ilaenv(3, "DGEHRD", " ", n, ilo, ihi, -1))
+		if nx < nh {
+			// Determine if workspace is large enough for blocked code.
+			if lwork < n*nb+tsize {
+				// Not enough workspace to use optimal nb:
+				// determine the minimum value of nb, and reduce
+				// nb or force use of unblocked code.
+				nbmin = max(2, impl.Ilaenv(2, "DGEHRD", " ", n, ilo, ihi, -1))
+				if lwork >= n*nbmin+tsize {
+					nb = (lwork - tsize) / n
+				} else {
+					nb = 1
+				}
+			}
+		}
+	}
+	ldwork := nb // work is used as an n×nb matrix.
+
+	var i int
+	if nb < nbmin || nh <= nb {
+		// Use unblocked code below.
+		i = ilo
+	} else {
+		// Use blocked code.
+
+		bi := blas64.Implementation()
+		iwt := n * nb // Size of the matrix Y and index where the matrix T starts in work.
+		for i = ilo; i < ihi-nx; i += nb {
```

Comment:
I noticed that @btracey in `dgebrd.go` changed `nx` in a very similar loop to `nb`. `nx` is the crossover block size between blocked and unblocked code, its default value is 128, while the default block size is 32 (minimum 2). This does seem a bit strange because the loop is ever started to be executed only for matrices larger than 129 (even if the minimum block size of 2 is used). I'm not sure what the right action is here, I'm just point it out. I would like to merge this PR so that I can continue submitting what I have in the queue.